toolchain/lex/lex.cpp

```diff
@@ -708,6 +720,27 @@ auto Lexer::LexVerticalWhitespace(llvm::StringRef source_text,
   line_info->indent = position - line_start;
 }
 
+auto Lexer::LexCR(llvm::StringRef source_text, ssize_t& position) -> void {
+  if (LLVM_LIKELY((position + 1) < static_cast<ssize_t>(source_text.size())) &&
+      LLVM_LIKELY(source_text[position + 1] == '\n')) {
+    // Skip to the vertical whitespace path, it will skip over both CR and LF.
+    LexVerticalWhitespace(source_text, position);
+    return;
+  }
+
+  CARBON_DIAGNOSTIC(UnsupportedLFCRLineEnding, Error,
+                    "The LF+CR line ending is not supported, only LF and CR+LF "
+                    "is supported.");
+  CARBON_DIAGNOSTIC(
+      UnsupportedCRLineEnding, Error,
+      "A raw CR line ending is not supported, only LF and CR+LF is supported.");
+  bool is_lfcr = position > 0 && source_text[position - 1] == '\n';
+  emitter_.Emit(source_text.begin() + position,
+                is_lfcr ? UnsupportedLFCRLineEnding : UnsupportedCRLineEnding);
```

Comment:
Do we do anything to prevent the CR in the snippet from garbling the diagnostic output?